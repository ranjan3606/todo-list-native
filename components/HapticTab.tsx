import React from 'react';
import { GestureResponderEvent, Pressable, Platform, StyleSheet } from 'react-native';
import * as Haptics from 'expo-haptics';
import { BottomTabBarButtonProps } from '@react-navigation/bottom-tabs';
import { useColorScheme } from 'react-native';
import { Colors } from '@/constants/Colors';

export interface HapticTabProps extends BottomTabBarButtonProps {
  onTabPress?: () => void;
}

export function HapticTab(props: HapticTabProps) {
  const { accessibilityState, onPress, children, onTabPress } = props;
  const colorScheme = useColorScheme() ?? 'light';
  const isSelected = accessibilityState?.selected;
  
  const handlePress = (event: GestureResponderEvent) => {
    if (Platform.OS === 'ios') {
      Haptics.impactAsync(Haptics.ImpactFeedbackStyle.Light);
    }
    
    // Call the onTabPress callback if provided
    if (onTabPress) {
      onTabPress();
    }
    
    // Call the original onPress with the event
    if (onPress) {
      onPress(event);
    }
  };

  return (
    <Pressable
      {...props}
      onPress={handlePress}
      role="button"
      style={({ pressed }) => [
        styles.tab,
        { 
          opacity: pressed ? 0.7 : 1,
          backgroundColor: isSelected ? 
            Colors[colorScheme].tabActiveBackground : 
            'transparent',
        }
      ]}
    >
      {children}
    </Pressable>
  );
}

const styles = StyleSheet.create({
  tab: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    height: 50,
    paddingVertical: 8,
    borderRadius: 8,
    marginHorizontal: 5,
  }
});
