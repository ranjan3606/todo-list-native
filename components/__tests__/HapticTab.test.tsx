import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { HapticTab, HapticTabProps } from '../HapticTab';
import * as Haptics from 'expo-haptics';
import { Platform } from 'react-native';

jest.mock('@react-native-async-storage/async-storage', () => ({
  getItem: jest.fn(() => Promise.resolve(null)),
  setItem: jest.fn(() => Promise.resolve(null)),
  removeItem: jest.fn(() => Promise.resolve(null)),
  clear: jest.fn(() => Promise.resolve(null)),
  getAllKeys: jest.fn(() => Promise.resolve([])),
  multiGet: jest.fn(() => Promise.resolve([])),
  multiSet: jest.fn(() => Promise.resolve(null)),
  multiRemove: jest.fn(() => Promise.resolve(null)),
}));

jest.mock('expo-haptics', () => ({
  impactAsync: jest.fn(),
  ImpactFeedbackStyle: { Light: 'light' }
}));

describe('HapticTab', () => {
  const defaultProps: HapticTabProps = {
    accessibilityState: { selected: false },
    onPress: jest.fn(),
    children: <></>,
  };
  
  // Store original Platform.OS
  const originalOS = Platform.OS;

  afterEach(() => {
    jest.clearAllMocks();
    // Restore original Platform.OS after each test
    Platform.OS = originalOS;
  });

  it('renders correctly (snapshot)', () => {
    const { toJSON } = render(<HapticTab {...defaultProps} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('applies selected style when selected', () => {
    const { getByRole } = render(
      <HapticTab {...defaultProps} accessibilityState={{ selected: true }} />
    );
    const pressable = getByRole('button');
    expect(pressable.props.style[1].backgroundColor).not.toBe('transparent');
  });

  it('calls onTabPress and onPress when pressed', () => {
    const onTabPress = jest.fn();
    const onPress = jest.fn();
    const { getByRole } = render(
      <HapticTab {...defaultProps} onTabPress={onTabPress} onPress={onPress} />
    );
    fireEvent.press(getByRole('button'));
    expect(onTabPress).toHaveBeenCalled();
    expect(onPress).toHaveBeenCalled();
  });

  it('triggers haptic feedback on iOS', () => {
    // Directly set Platform.OS instead of using spyOn
    Platform.OS = 'ios';
    const { getByRole } = render(<HapticTab {...defaultProps} />);
    fireEvent.press(getByRole('button'));
    expect(Haptics.impactAsync).toHaveBeenCalledWith(Haptics.ImpactFeedbackStyle.Light);
  });

  it('does not trigger haptic feedback on Android', () => {
    // Directly set Platform.OS instead of using spyOn
    Platform.OS = 'android';
    const { getByRole } = render(<HapticTab {...defaultProps} />);
    fireEvent.press(getByRole('button'));
    expect(Haptics.impactAsync).not.toHaveBeenCalled();
  });
});
