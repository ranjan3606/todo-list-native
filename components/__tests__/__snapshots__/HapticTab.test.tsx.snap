// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HapticTab renders correctly (snapshot) 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": false,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  role="button"
  style={
    [
      {
        "alignItems": "center",
        "borderRadius": 8,
        "flex": 1,
        "height": 50,
        "justifyContent": "center",
        "marginHorizontal": 5,
        "paddingVertical": 8,
      },
      {
        "backgroundColor": "transparent",
        "opacity": 1,
      },
    ]
  }
/>
`;
